File: internal/controllers/volume.go
// Copyright 2025 The wharf Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controllers

import (
	"context"
	"log"
	"net/http"
	"time"

	"github.com/docker/docker/api/types/volume"
	"github.com/docker/docker/client"
	"github.com/docker/docker/errdefs"
	"github.com/gin-gonic/gin"
	"github.com/go-playground/validator/v10"

	"github.com/wharf/wharf/pkg/models"
	dockerVolume "github.com/wharf/wharf/pkg/volume"
)

func GetVolumes(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		volumes, err := dockerVolume.GetAll(ctx, dockerClient)
		if err != nil {
			log.Println(err)
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(200, volumes)
	}
}

func RemoveVolume(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		id := c.Param("id")
		var reqBody dockerVolume.RemoveVolumeRequest
		ur, _ := c.Get("user")
		reqUser, _ := ur.(models.User)
		if reqUser.Permission == models.Read {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid permissions"})
			return
		}
		if err := c.BindJSON(&reqBody); err != nil {
			log.Println(err)
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}

		forceRem := false
		if reqBody.Force != nil {
			forceRem = *reqBody.Force
		}
		if err := dockerVolume.Remove(ctx, dockerClient, id, forceRem); err != nil {
			if errdefs.IsNotFound(err) {
				c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
				return
			}
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, gin.H{"message": id + " volume removed"})

	}

}

func PruneVolumes(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		var ctx, cancel = context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		ur, _ := c.Get("user")
		reqUser, _ := ur.(models.User)

		if reqUser.Permission != models.Execute {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid permissions"})
			return
		}

		report, err := dockerVolume.Prune(ctx, dockerClient)
		if err != nil {
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, report)
	}
}

func CreateVolume(dockerClient *client.Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()
		var createVolumeRequest dockerVolume.CreateVolumeRequest
		ur, _ := c.Get("user")
		reqUser, _ := ur.(models.User)
		if reqUser.Permission == models.Read {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid permissions"})
			return
		}
		if err := c.BindJSON(&createVolumeRequest); err != nil {
			log.Println(err)
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}
		validate := validator.New()
		if err := validate.Struct(createVolumeRequest); err != nil {
			log.Println(err)
			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
			return
		}
		opts := volume.CreateOptions{
			Name:   createVolumeRequest.Name,
			Driver: "local",
		}

		vol, err := dockerVolume.Create(ctx, dockerClient, opts)
		if err != nil {
			c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
			return
		}
		c.JSON(http.StatusOK, vol)
	}
}

File: client/src/api/volume.ts
// Copyright 2025 The wharf Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import axios, { AxiosResponse } from 'axios';
import { hostUrl } from '../utils/util';
import { Volume, VolumesPruneReport } from '../models/volume';
import { MessageRes } from '../models/common';

const URL = hostUrl();

export const getAllVolumes = async (
  token: string
): Promise<AxiosResponse<Volume[]>> => {
  const url = URL + '/api/protected/volume/getAll';
  const res = await axios.get<Volume[]>(url, {
    headers: {
      Token: token,
    },
  });
  return res;
};

export const deleteVolume = async (
  token: string,
  id: string,
  force: boolean
): Promise<AxiosResponse<MessageRes>> => {
  const url = URL + '/api/protected/volume/remove/' + id;
  const res = await axios.delete<MessageRes>(url, {
    data: {
      force: force,
    },
    headers: {
      Token: token,
    },
  });
  return res;
};

export const pruneVolumes = async (
  token: string
): Promise<AxiosResponse<VolumesPruneReport>> => {
  const url = URL + '/api/protected/volume/prune';
  const res = await axios.delete<VolumesPruneReport>(url, {
    headers: {
      Token: token,
    },
  });
  return res;
};

export const createVolume = async (
  token: string,
  name: string
): Promise<AxiosResponse<Volume>> => {
  const url = URL + '/api/protected/volume/create';
  const res = await axios.post<Volume>(
    url,
    {
      name: name,
    },
    {
      headers: {
        Token: token,
      },
    }
  );
  return res;
};
